fictadvisor-api/src/v2/api/services/CathedraService.ts

```diff
@@ -3,37 +3,105 @@ import { CathedraRepository } from '../../database/repositories/CathedraReposito
 import { CreateCathedraDTO } from '../dtos/CreateCathedraDTO';
 import { UpdateCathedraDTO } from '../dtos/UpdateCathedraDTO';
 import { DbCathedra } from '../../database/entities/DbCathedra';
-import { QueryAllDTO } from '../../utils/QueryAllDTO';
 import { DatabaseUtils } from '../../database/DatabaseUtils';
+import { QueryAllCathedrasDTO, SortQACParam } from '../dtos/QueryAllCathedrasDTO';
+import { CathedraMapper } from '../../mappers/CathedraMapper';
 
 @Injectable()
 export class CathedraService {
   constructor (
     private cathedraRepository: CathedraRepository,
+    private cathedraMapper: CathedraMapper
   ) {}
 
-  async create (body: CreateCathedraDTO): Promise<DbCathedra> {
-    return this.cathedraRepository.create(body);
+  create (body: CreateCathedraDTO): Promise<DbCathedra> {
+    return this.cathedraRepository.create({
+      ...body,
+      teachers: {
+        createMany: {
+          data: body.teachers?.map((teacherId) => ({ teacherId })) || [],
+        },
+      },
+    });
   }
 
   async update (id: string, body: UpdateCathedraDTO): Promise<DbCathedra>  {
-    return this.cathedraRepository.updateById(id, body);
+    const { addTeachers, deleteTeachers } = body;
+
+    return this.cathedraRepository.updateById(id, {
+      name: body.name,
+      abbreviation: body.abbreviation,
+      division: body.division,
+      teachers: {
+        createMany: {
+          data: addTeachers?.map((teacherId) => ({ teacherId })) || [],
+        },
+        deleteMany: {
+          teacherId: {
+            in: deleteTeachers || [],
+          },
+        },
+      },
+    });
   }
 
   async delete (id: string): Promise<DbCathedra> {
     return this.cathedraRepository.deleteById(id);
   }
 
-  async getAll (body: QueryAllDTO) {
-    const search = DatabaseUtils.getSearch(body, 'name', 'abbreviation');
-    const sort = DatabaseUtils.getSort(body, 'name');
+  async getAll (query: QueryAllCathedrasDTO) {
+    const search = {
+      AND: [
+        this.getCathedraSearch.name(query.search),
+        this.getCathedraSearch.abbreviation(query.abbreviation),
+        this.getCathedraSearch.divisions(query.divisions),
+      ],
+    };
+
+    const sort = query.sort === SortQACParam.TEACHERS ? {
+      orderBy: {
+        teachers: {
+          _count: query?.order || 'asc',
+        },
+      },
+    } : DatabaseUtils.getSort(query, SortQACParam.NAME);
 
     const data = {
       ...sort,
+      where: search,
+    };
+
+    return await DatabaseUtils.paginate<DbCathedra>(this.cathedraRepository, query, data);
+  }
+
+  private getCathedraSearch = {
+    name: (search: string) => DatabaseUtils.getSearch({ search }, 'name'),
+    abbreviation: (search: string) => DatabaseUtils.getSearch({ search }, 'abbreviation'),
+    divisions: (divisions: string[]) => DatabaseUtils.getSearchByArray(divisions, 'division'),
+  };
+
+  async getCathedras (cathedraId: string, notInDepartment: boolean): Promise<DbCathedra[]> {
```

Comment:
* Rename `getCathedras` to `getTeachersFromCathedra`
* Make `notInDepartment` optional